Generate jni bindings

midepeter opened the pull request:
This PR is to realated to initial !1327 for generating java bindings for the ach library

midepeter committed:
'Mergefiles' improvements

This commit gives a benchmark report on the use of '[]*ach.File' and '[]ach.File'. It intend to show the improvement in passing by values instead of passing by reference

midepeter committed:
java bindings exceptions generation

this commit creates javainterop package that ports functions from the general moov/ach package that does not align with the gobind rules for java jni bindings generation. It does this to be able to release a proper generated code for java apis to all the ach library.

midepeter committed:
clean up

adamdecaf commented on javainterop/nextEntry.go:
```diff
@@ -0,0 +1,29 @@
+package javainterop
```
Can we put all of this in one `javainterop.go` file that has an alias for every type? We can figure out a script or something to generate the file if that's easier.

midepeter replied:
that javainterop.go should have an alias for every type in the ach package right

are we porting every function in the ach package to the javainterop package or just the ones we are trying to reformat (SegmentFile and NextEntry)

adamdecaf replied:
Yes it should have an alias to every type and function so they're available to someone using Java. We need to offer wrappers for some of the functions that don't compile nicely.

midepeter replied:
Oh alright.

midepeter replied:
@adamdecaf 

I am having a `infinite recursive call` lint issue when calling the `ach` functions in the `javainterop` functions. I don't think this affects the gobind generation but should i be mindful of that?

adamdecaf replied:
Can you share the full error?

midepeter replied:
`infinite recursive call (SA5007)go-staticcheck`

It is not a full error. Just a warning for the go-staticcheck

adamdecaf replied:
Are you trying to generate the JNI code for the entire moov-io/ach repository? You should be able to target the `javainterop` package as the only thing to compile/generate.

midepeter replied:
No, I am targeting just the `javainterop` package.

Actually, I realized if we go the route of referencing the ach types by embedding them into structs and calling the exported functions on them like you highlighted for the `Iterator` and `NextEntry` below. It eliminates this error. I think it is because of the `alias` as it is, does not necessarily create a newType but just masks the `oldType`

adamdecaf replied:
Ah that could be. Sounds good!

adamdecaf commented on javainterop/nextEntry.go:
```diff
@@ -0,0 +1,29 @@
+		NextBatchHeader: BatchHeader(*batchHeader),
+		NextEntryDetail: EntryDetail(*entryFile),
+		NextEntryError:  err,
+	}
```
We will want to make the `Iterator` a struct holding a reference to the underlying `ach.Iterator` so that each call to `NextEntry()` will progress record by record. What you have now wouldn't quite work.

midepeter replied:
Yes, this is true.

midepeter committed:
porting exported types to a single javainterop package

This commit ports exported types and methods to a single javainterop package to allow for ease generation of go bindings to allow for java application to be largely supported. It also updates the github actions workflow to generate the binding on push to master branches by access the 'make java' command in the makeFile

midepeter committed:
Merge remote-tracking branch 'origin/master' into generate-jni-bindings

midepeter committed:
chore: linter fix

midepeter committed:
linterissue fix

midepeter committed:
tidy up go modules
